papaya-web/src/main/java/pl/edu/agh/papaya/notification/NotificationService.java

```diff
@@ -0,0 +1,162 @@
+import java.util.stream.Collectors;
+import lombok.RequiredArgsConstructor;
+import org.springframework.mail.javamail.JavaMailSender;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Service;
+import pl.edu.agh.papaya.model.Notification;
+import pl.edu.agh.papaya.model.NotificationType;
+import pl.edu.agh.papaya.model.Sprint;
+import pl.edu.agh.papaya.model.SprintState;
+import pl.edu.agh.papaya.model.UserInProject;
+import pl.edu.agh.papaya.notification.message.EmailMessage;
+import pl.edu.agh.papaya.notification.message.Message;
+import pl.edu.agh.papaya.notification.message.WebhookMessage;
+import pl.edu.agh.papaya.repository.AvailabilityRepository;
+import pl.edu.agh.papaya.repository.NotificationRepository;
+import pl.edu.agh.papaya.service.sprint.SprintService;
+import pl.edu.agh.papaya.service.userinproject.UserInProjectService;
+
+@Service
+@RequiredArgsConstructor
+@SuppressWarnings({"PMD.BeanMembersShouldSerialize"})
+public class NotificationService {
+
+    // TODO: Set this to something reasonable
+    private static final Duration NOTIFICATION_FREQUENCY = Duration.ofMinutes(1);
+
+    private final SprintService sprintService;
+
+    private final UserInProjectService userInProjectService;
+
+    private final NotificationRepository notificationRepository;
+
+    private final AvailabilityRepository availabilityRepository;
+
+    private final JavaMailSender javaMailSender;
+
+    // TODO: Set this to something reasonable
+    @Scheduled(cron = "0/20 * * * * ?")
+    public void sendNotifications() {
+        LocalDateTime currentTime = LocalDateTime.now();
+        sendDeclarationsOpenOrReminderNotifications(currentTime);
+        sendDeclarationsFinishedNotifications(currentTime);
+    }
+
+    private void sendDeclarationsFinishedNotifications(LocalDateTime currentTime) {
+        List<Sprint> paddingSprints = sprintService.findByState(SprintState.PADDING, currentTime);
+
+        paddingSprints.forEach(sprint -> {
+            var project = sprint.getProject();
+            var usersInProjects = userInProjectService.findActiveUsersByProject(project);
+            var notificationStrategy = NotificationStrategy.resolveNotificationStrategy(project);
+
+            Optional<Notification> lastNotification = Optional.ofNullable(
+                    Iterables.getLast(notificationRepository.findSprintNotifications(sprint), null));
+
+            if (lastNotification.isPresent() && sprintClosedMessageNotYetSent(lastNotification.get())) {
+                Optional<Message> message = createSprintClosedMessage(sprint, usersInProjects, notificationStrategy);
+                message.ifPresent(value -> sendAndUpdate(currentTime, sprint, value));
+            }
+        });
+    }
+
+    private void sendAndUpdate(LocalDateTime currentTime, Sprint sprint, Message message) {
+        try {
+            message.send();
+        } catch (NotificationMessageException e) {
+            return;
```

Comment:
W sumie nie wiem, co z nim tutaj ładnie zrobić

Reply:
Na devie to się będzie ciągle krzaczyć, bo jest "wyłączone" przy pomocy maila nadawcy `@example.com` i jakiegoś randomowego hasła, więc nigdy nie przejdzie, chyba że będzie valid webhook

Reply:
może properties emailNotificationsEnable wyłączony na devie?

Reply:
Dev dostał crona na "nigdy"

Chodziło mi bardziej o to, że to chyba nie jest ładne:
```
try {
        cośTam();
 } catch (NotificationMessageException e) {
        return;
}
cośTam();
```